src/mongo/db/s/shard_filtering_metadata_refresh.cpp

```diff
@@ -376,25 +376,51 @@ SharedSemiFuture<void> recoverRefreshShardVersion(ServiceContext* serviceContext
 
             ON_BLOCK_EXIT([&] {
+                boost::optional<SharedSemiFuture<void>> waitForMigrationAbort;
                 UninterruptibleLockGuard noInterrupt(opCtx->lockState());
-                // A view can potentially be created after spawning a thread to recover nss's shard
-                // version. It is then ok to lock views in order to clear filtering metadata.
-                //
-                // DBLock and CollectionLock must be used in order to avoid shard version checks
-                Lock::DBLock dbLock(opCtx, nss.db(), MODE_IX);
-                Lock::CollectionLock collLock(opCtx, nss, MODE_IX);
-
-                auto* const csr = CollectionShardingRuntime::get(opCtx, nss);
-
-                auto csrLock = CollectionShardingRuntime::CSRLock::lockExclusive(opCtx, csr);
-                // cancellationToken needs to be checked under the CSR lock before overwriting the
-                // filtering metadata to serialize with other threads calling
-                // 'clearFilteringMetadata'
-                if (currentMetadataToInstall && !cancellationToken.isCanceled()) {
-                    csr->setFilteringMetadata_withLock(opCtx, *currentMetadataToInstall, csrLock);
+                {
+                    // A view can potentially be created after spawning a thread to recover nss's
+                    // shard version. It is then ok to lock views in order to clear filtering
+                    // metadata.
+                    //
+                    // DBLock and CollectionLock must be used in order to avoid shard version checks
+                    Lock::DBLock dbLock(opCtx, nss.db(), MODE_IX);
+                    Lock::CollectionLock collLock(opCtx, nss, MODE_IX);
+
+                    auto* const csr = CollectionShardingRuntime::get(opCtx, nss);
+
+                    auto csrLock = CollectionShardingRuntime::CSRLock::lockExclusive(opCtx, csr);
+                    // cancellationToken needs to be checked under the CSR lock before overwriting
+                    // the filtering metadata to serialize with other threads calling
+                    // 'clearFilteringMetadata'
+                    if (currentMetadataToInstall && !cancellationToken.isCanceled()) {
+                        csr->setFilteringMetadata_withLock(
+                            opCtx, *currentMetadataToInstall, csrLock);
+
+                        if (currentMetadataToInstall->isSharded() &&
+                            !currentMetadataToInstall->allowMigrations()) {
+                            if (auto msm = MigrationSourceManager::get(csr, csrLock)) {
+                                waitForMigrationAbort.emplace(msm->abort());
+                            }
+                        }
+                    }
+                }
+
+                // Join any ongoing migration outside of the CSR lock.
+                if (waitForMigrationAbort) {
+                    waitForMigrationAbort->get(opCtx);
```

Comment:
I can't remember what is the semantics of UninterruptibleLockGuard, but I thought it only applies to locks. If so, then this here can throw and will crash the process because ON_BLOCK_EXIT is really a destructor.

If it doesn't apply only to locks, then it does look a little bit sketchy from a potential deadlock perspective, since there is a refresh in the MSM. I think it is fine because the [refresh in the MSM](https://github.com/mongodb/mongo/blob/eafdc7cc185d4dac70f61fad5ffbbe54bfa67e31/src/mongo/db/s/migration_source_manager.cpp#L199) happens before it is [registered](https://github.com/mongodb/mongo/blob/eafdc7cc185d4dac70f61fad5ffbbe54bfa67e31/src/mongo/db/s/migration_source_manager.cpp#L254) therefore we can never have a circular wait.

Can you please confirm which one is the case and if it is the latter, please add a comment in both places indicating this dependency.

Reply:
I checked and like you said, it only affects locks, so, I've changed the code to only have the locks uninterruptible.

Reply:
But then don't we still have the first problem - where this can throw and because ON_BLOCK_EXIT is a destructor, it will result in `std::terminate`? And you can't just getNoExcept/catch the exception because then we will violate the contract that if `allowMigrations` has been set by `onShardVersionMismatch` we will have joined the migrations when it returns 😕 

So I am proposing that we just get without a deadline so it can't throw, which would be equivalent to the latter scenario I described.